### DISCUSSION TITLE
Bitcoin key type system

### DISCUSSION THREAD

**Discussion by dr-orlovsky** [rust-bitcoin/rust-bitcoin#919] [Refactoring]
Some time ago we had a lot of discussions about the best design for bitcoin primitive type system (#550). One of the most discussed parts for long time was refactoring of public key system. Some of the previous discussion include:
- in context of new taproot keys: https://github.com/rust-bitcoin/rust-bitcoin/issues/588
- early discussions: https://github.com/rust-bitcoin/rust-bitcoin/issues/382

I tried to imagine a key API & type system I'd like to have, taking as an example public key types. This is a brainstorming attempt of a sort "what if I had to design bitcoin key types from the scratch, optimizing for performance and optionality of secp256k1 dependence".

```rust

pub mod field {
    pub struct Value([u8; 32]);
    // impl Add, Sub, CheckedAdd, WrappedAdd, CheckedSub, WrappedSub for Value
}

pub mod ec {
    use super::field;

    pub trait Point
    where
        Self: Add + Sub + Mul + From<Self::Legacy>,
        Self::Legacy: From<Self>,
    {
        type LegacyType; // like bitcoin::PublicKey, XOnlyPublicKey etc
        fn x(&self) -> field::Value;
        fn y(&self) -> field::Value;
        fn parity(&self) -> Parity;
        fn check_ec_point(&self) -> Result<(), NotCurvePoint>;
        fn to_secp_pk(&self) -> secp256k1::PublicKey;
        fn to_legacy(&self) -> Self::LegacyType;
    }

    pub enum Parity {
        Even,
        Odd,
    }
}

pub mod key {
    use super::{ec, field};
    use std::fmt::{self, Formatter};

    pub trait Secret {
        fn fmt_secret(&self, f: Formatter) -> fmt::Result<()>;
        fn display_secret(&self) -> DisplaySecret;
        fn to_string_secret(&self) -> String;
    }

    pub struct Seckey(field::Value);
    // impl Secret for Seckey

    pub struct UncompressedPubkey {
        x: field::Value,
        y: field::Value,
    }
    // impl Point for UncompressedPubkey

    pub struct CompressedPubkey {
        parity: ec::PointParity,
        x: field::Value,
    }
    // impl Point for CompressedPubkey

    pub struct XOnlyPubkey {
        x: field::Value,
    }
    // impl Point for XOnlyPubkey

    pub enum LegacyPubkey {
        Uncompressed(UncompressedPubkey),
        Compressed(CompressedPubkey),
    }
    // impl Point for LegacyPubkey

    pub type SegwitV0Pubkey = CompressedPubkey;
    pub type InternalPubkey = XOnlyPubkey;
    pub struct OutputPubkey(XOnlyPubkey);
    // impl Point for OutputPubkey

    pub struct KeyPair {
        seckey: Seckey,
        pubkey: XOnlyPubkey,
    }
    // impl Point for KeyPair
    // impl Secret for KeyPair
}
```

**Comment by apoelstra** [comment]
The fielrd/point stuff should not be public. They are too low-level for libsecp and definitely too low-level for this library.

I also think that calling scalar values a "field" is confusing. Yes, these values do accidentally form a field since they are isomorphic to the integers modulo a prime; but this is not true for all EC groups. In libsecp we use the term "scalar" which we should continue to use here (if this scalar/point distinction is something we want to do at all, which I'm unconvinced of).

>optionality of secp256k1 dependence"

Absent libsecp the only functionality that I would be comfortable retaining would be de/serialization (including hashing) of keys. I'm not sure if this is what you're going for here, but I think pretty-much any API would be compatible with that.

**Reply by dr-orlovsky** [reply]
The idea is to feature-gate all "write" methods operating keys and link to the presence of secp lib. If the lib is present, this is full-fledged implementation, not only for parsing. If the lib is not linked via feature flag, it is just a parsing library.

Methods operating real keys will look like

```rust
#[cfg(feature = "secp")]
impl TapTweak for XOnlyPubkey {
    fn tap_tweak(&self, tweak: TapTweak) -> Result<TweakedPubkey, Error> {
         let pk = secp256k1::XOnlyPublicKey::from_bytes(self.into_bytes())?;
         // or
         let pk: secp256k1::XOnlyPublicKey = self.try_into()?;
         pk.add_assign(tweak.into_bytes()).map_err(....)
    }
}

```

This will make the code which is not required to validate the keys much faster (not calling each time to secp library just for parsing etc). Also the data inside pubkeys will be platform-independent.

> I also think that calling scalar values a "field" is confusing.

We can probably name them differently (as well as `ec`). The idea is to keep just byte data, without crypto processing unless it is needed - and we need some 32-byte inner type to standardize most of the byte access operations across different key types.

**Reply by Kixunil** [reply]
I'd prefer to postpone bikeshedding but want to not forget I also don't like `field` because it can be confused with struct field. `scalar` sounds good.

**Comment by apoelstra** [comment]
I really dislike `secp256k1::XOnlyPublicKey::from_bytes(self.into_bytes())?`. The implication seems to be that we would allow invalid key data until the time it was actually used?

**Reply by dr-orlovsky** [reply]
The same we allow invalid scripts not verifying them with `libbitcoinconsensus` when parsing, constructing etc.

**Reply by apoelstra** [reply]
There are no "invalid scripts", every bytestring is a valid script.

Absolute NACK to changing our keytypes to allow representing things that aren't even keys.

**Comment by Kixunil** [comment]
Interesting! Things I like:

* `CompressedPubkey` 
* The `Point` trait looks useful
* Struct fields are private
* The idea of making libsecp256k1 optional; though wouldn't we have to re-implement some things in Rust to do proper parsing? I guess at least checking if the point is on the curve.
* `Scalar` (`Value`) type looks useful.

Things I'm not sure about:

* Is `LegacyPubkey` being an enum and an additional type really better?
* How useful are the type aliases (vs newtypes)
* Naming - bikeshedding to do later
* `check_ec_point()` - the types should guarantee being on the curve but then we probably can't construct them without `libsecp256k1` or re-impl.
* The legacy conversions - either the trait should be on top of the existing types or we just break backcompat if required.

**Reply by dr-orlovsky** [reply]
> The idea of making libsecp256k1 optional; though wouldn't we have to re-implement some things in Rust to do proper parsing? I guess at least checking if the point is on the curve.

I think this is a great idea which will solve a lot of issues here: to re-implement a single operation from libsecp256k1 (the check that the key is on the curve; plus trivial check that field member is within the field for secret keys).

> The legacy conversions - either the trait should be on top of the existing types or we just break backcompat if required.

I am thinking of starting a separate creates doing all these types from scratch (`bitcoin_keys` and `bitcoin_scripts`, for the companion discussion) - and then migrating current `rust-bitcoin` on it, as a part of `rust-bitcoin` multicrate refactoring.

**Reply by Kixunil** [reply]
Trying in a separate crate sounds good.

**Comment by apoelstra** [comment]
What about implementing parsing/keycheck in rust-secp, and adding a feature to enable/disable the C library?

**Reply by Kixunil** [reply]
Yeah, I had that idea as well. Thinking about it now it seems clearly better - other projects may want to use it too and it will keep the types same. I don't really see any downside.

**Reply by dr-orlovsky** [reply]
That was the thing I tried to propose here, sorry for not doing that explicitly enough. That's why I had feature gates on the secp-related key methods

**Reply by apoelstra** [reply]
Ok, I'll explore doing this in rust-secp.
